nexus-sdk/src/main/java/io/nexusrpc/Serializer.java

```diff
@@ -0,0 +1,99 @@
+package io.nexusrpc;
+
+import java.lang.reflect.Type;
+import java.util.*;
+import org.jspecify.annotations.Nullable;
+
+/** Serializer used to convert values to bytes and vice-versa. */
+public interface Serializer {
+  /** Serialize the value as content. */
+  Content serialize(@Nullable Object value);
+
+  /** Deserialize the content as a value of the given type. */
+  @Nullable Object deserialize(Content content, Type type);
+
+  /** Data + headers used by serializers. */
+  class Content {
+    /** Create a builder for content. */
+    public static Builder newBuilder() {
+      return new Builder();
+    }
+
+    /** Create a builder for content from existing content. */
+    public static Builder newBuilder(Content content) {
+      return new Builder(content);
+    }
+
+    // TODO(cretz): Is this acceptable over ByteBuffer for our use case?
+    private final byte[] data;
+    private final Map<String, String> headers;
+
+    private Content(byte[] data, Map<String, String> headers) {
+      this.data = data;
+      this.headers = headers;
+    }
+
+    /** Data. */
+    public byte[] getData() {
+      return data;
+    }
+
+    /** Headers. */
+    public Map<String, String> getHeaders() {
+      return headers;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      Content content = (Content) o;
+      return Objects.deepEquals(data, content.data) && Objects.equals(headers, content.headers);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(Arrays.hashCode(data), headers);
+    }
+
+    /** Builder for content. */
+    public static class Builder {
+      private byte @Nullable [] data;
+      private final Map<String, String> headers;
+
+      private Builder() {
+        headers = new HashMap<>();
+      }
+
+      private Builder(Content content) {
+        data = content.data;
+        headers = new HashMap<>(content.headers);
+      }
+
+      /** Set data. Required. */
+      public Builder setData(byte[] data) {
+        this.data = data;
+        return this;
+      }
+
+      /** Get headers to mutate. */
+      public Map<String, String> getHeaders() {
+        return headers;
+      }
+
+      /** Add header. */
+      public Builder putHeader(String key, String value) {
```

Comment:
What headers are these supposed to be? The Neuxs Go SDK has a concept of `LazyValue`, but those have not header so I can't figure out what these headers map to

Reply:
This is the same as https://pkg.go.dev/github.com/nexus-rpc/sdk-go/nexus#Content `Header` (and this class is named `Content` for that reason). I was afraid to move this to a top-level class because it seemed very serializer specific.